File: libnd4j/include/ops/declarable/generic/nn/loha_matmul.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// LoHa (Low-Rank Hadamard Product) fused matrix multiplication operation.
// Computes: output = input @ weight^T + scaling * input @ ((B1 @ A1) ⊙ (B2 @ A2))^T
//
// @author Adam Gibson
//

#include <system/op_boilerplate.h>
#include <array/NDArrayFactory.h>
#if NOT_EXCLUDED(OP_loha_matmul)

#include <ops/declarable/headers/parity_ops.h>
#include <ops/declarable/headers/blas.h>
#include <ops/declarable/helpers/matmul.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(loha_matmul, 6, 1, false, 0, 0) {
  auto input = INPUT_VARIABLE(0);   // [batch, in_features]
  auto weight = INPUT_VARIABLE(1);  // [out_features, in_features]
  auto lohaA1 = INPUT_VARIABLE(2);  // [dim, in_features]
  auto lohaB1 = INPUT_VARIABLE(3);  // [out_features, dim]
  auto lohaA2 = INPUT_VARIABLE(4);  // [dim, in_features]
  auto lohaB2 = INPUT_VARIABLE(5);  // [out_features, dim]
  auto output = OUTPUT_VARIABLE(0); // [batch, out_features]

  if (input->isEmpty() || weight->isEmpty()) {
    return Status::OK;
  }

  double scaling = block.getTArguments()->size() > 0 ? T_ARG(0) : 1.0;
  bool transposeWeight = block.getBArguments()->size() > 0 ? B_ARG(0) : true;

  REQUIRE_TRUE(input->rankOf() == 2, 0,
               "loha_matmul: Input should have rank 2, got %i", input->rankOf());

  // Step 1: Compute base output: input @ weight^T
  auto baseOutput = output->ulike();  // returns NDArray*
  {
    sd::ops::matmul mmulOp;
    std::vector<double> tArgs;
    std::vector<sd::LongType> iArgs = {0, transposeWeight ? 1 : 0};
    std::vector<bool> bArgs;
    std::vector<NDArray*> inputs = {input, weight};
    std::vector<NDArray*> outputs = {baseOutput};
    mmulOp.execute(inputs, outputs, tArgs, iArgs, bArgs);
  }

  // Step 2: Compute LoHa products
  auto outFeatures = weight->sizeAt(0);
  auto inFeatures = weight->sizeAt(1);

  std::vector<sd::LongType> prodShape = {outFeatures, inFeatures};
  auto prod1 = NDArrayFactory::create<float>('c', prodShape, input->getContext());
  auto prod2 = NDArrayFactory::create<float>('c', prodShape, input->getContext());

  {
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {lohaB1, lohaA1};
    std::vector<NDArray*> outputs = {prod1};
    mmulOp.execute(inputs, outputs);
  }
  {
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {lohaB2, lohaA2};
    std::vector<NDArray*> outputs = {prod2};
    mmulOp.execute(inputs, outputs);
  }

  // Step 3: Hadamard product: lohaDelta = prod1 * prod2
  auto lohaDelta = new NDArray(*prod1 * *prod2);

  // Step 4: Compute LoHa output: input @ lohaDelta^T
  auto lohaOutput = output->ulike();  // returns NDArray*
  {
    sd::ops::matmul mmulOp;
    std::vector<double> tArgs;
    std::vector<sd::LongType> iArgs = {0, 1};
    std::vector<bool> bArgs;
    std::vector<NDArray*> inputs = {input, lohaDelta};
    std::vector<NDArray*> outputs = {lohaOutput};
    mmulOp.execute(inputs, outputs, tArgs, iArgs, bArgs);
  }

  // Step 5: Combine: output = base_output + scaling * loha_output
  if (std::abs(scaling - 1.0) < 1e-9) {
    output->assign(*baseOutput + *lohaOutput);
  } else {
    lohaOutput->applyScalar(scalar::Multiply, scaling, lohaOutput);
    output->assign(*baseOutput + *lohaOutput);
  }

  delete prod1;
  delete prod2;
  delete lohaDelta;
  delete baseOutput;
  delete lohaOutput;

  return Status::OK;
}

DECLARE_SHAPE_FN(loha_matmul) {
  auto inShapeInfo = inputShape->at(0);
  auto wShapeInfo = inputShape->at(1);

  bool transposeWeight = block.getBArguments()->size() > 0 ? B_ARG(0) : true;

  sd::LongType batch = shape::sizeAt(inShapeInfo, static_cast<sd::LongType>(0));
  sd::LongType outFeatures = transposeWeight ?
      shape::sizeAt(wShapeInfo, static_cast<sd::LongType>(0)) :
      shape::sizeAt(wShapeInfo, static_cast<sd::LongType>(1));

  std::vector<sd::LongType> outputShape = {batch, outFeatures};
  auto outputShapeInfo = ConstantShapeHelper::getInstance().createShapeInfo(
      ArrayOptions::dataType(inShapeInfo), 'c', outputShape);

  return SHAPELIST(outputShapeInfo);
}

DECLARE_TYPES(loha_matmul) {
  getOpDescriptor()->setAllowedInputTypes(ANY)->setAllowedOutputTypes({ALL_FLOATS});
}

CUSTOM_OP_IMPL(loha_matmul_bp, 7, 6, false, 0, 0) {
  auto input = INPUT_VARIABLE(0);   // [batch, in_features]
  auto weight = INPUT_VARIABLE(1);  // [out_features, in_features]
  auto lohaA1 = INPUT_VARIABLE(2);  // [dim, in_features]
  auto lohaB1 = INPUT_VARIABLE(3);  // [out_features, dim]
  auto lohaA2 = INPUT_VARIABLE(4);  // [dim, in_features]
  auto lohaB2 = INPUT_VARIABLE(5);  // [out_features, dim]
  auto dLdOut = INPUT_VARIABLE(6);  // [batch, out_features]

  auto dLdInput = OUTPUT_VARIABLE(0);
  auto dLdWeight = OUTPUT_VARIABLE(1);
  auto dLdLohaA1 = OUTPUT_VARIABLE(2);
  auto dLdLohaB1 = OUTPUT_VARIABLE(3);
  auto dLdLohaA2 = OUTPUT_VARIABLE(4);
  auto dLdLohaB2 = OUTPUT_VARIABLE(5);

  if (input->isEmpty() || dLdOut->isEmpty()) {
    return Status::OK;
  }

  double scaling = block.getTArguments()->size() > 0 ? T_ARG(0) : 1.0;
  bool transposeWeight = block.getBArguments()->size() > 0 ? B_ARG(0) : true;

  // Gradient w.r.t. weight is zero (frozen)
  double zero = 0.0;
  dLdWeight->assign(zero);

  // Compute products
  auto outFeatures = weight->sizeAt(0);
  auto inFeatures = weight->sizeAt(1);

  std::vector<sd::LongType> prodShape = {outFeatures, inFeatures};
  auto prod1 = NDArrayFactory::create<float>('c', prodShape, input->getContext());
  auto prod2 = NDArrayFactory::create<float>('c', prodShape, input->getContext());

  {
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {lohaB1, lohaA1};
    std::vector<NDArray*> outputs = {prod1};
    mmulOp.execute(inputs, outputs);
  }
  {
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {lohaB2, lohaA2};
    std::vector<NDArray*> outputs = {prod2};
    mmulOp.execute(inputs, outputs);
  }

  auto lohaDelta = new NDArray(*prod1 * *prod2);

  // Gradient w.r.t. input
  {
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {dLdOut, weight};
    std::vector<NDArray*> outputs = {dLdInput};
    mmulOp.execute(inputs, outputs);
  }
  auto dLdInputLoha = dLdInput->ulike();
  {
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {dLdOut, lohaDelta};
    std::vector<NDArray*> outputs = {dLdInputLoha};
    mmulOp.execute(inputs, outputs);
  }
  dLdInputLoha->applyScalar(scalar::Multiply, scaling, dLdInputLoha);
  *dLdInput += *dLdInputLoha;
  delete dLdInputLoha;

  // Gradient w.r.t. LoHa components
  auto dLdOutTimesInput = NDArrayFactory::create<float>('c', prodShape, input->getContext());
  {
    auto dLdOutT = new NDArray(dLdOut->transpose());
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {dLdOutT, input};
    std::vector<NDArray*> outputs = {dLdOutTimesInput};
    mmulOp.execute(inputs, outputs);
    delete dLdOutT;
  }

  auto gradProd1 = new NDArray(*dLdOutTimesInput * *prod2);
  gradProd1->applyScalar(scalar::Multiply, scaling, gradProd1);
  auto gradProd2 = new NDArray(*dLdOutTimesInput * *prod1);
  gradProd2->applyScalar(scalar::Multiply, scaling, gradProd2);

  // dLdA1 = B1^T @ gradProd1
  {
    auto lohaB1T = new NDArray(lohaB1->transpose());
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {lohaB1T, gradProd1};
    std::vector<NDArray*> outputs = {dLdLohaA1};
    mmulOp.execute(inputs, outputs);
    delete lohaB1T;
  }

  // dLdB1 = gradProd1 @ A1^T
  {
    auto lohaA1T = new NDArray(lohaA1->transpose());
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {gradProd1, lohaA1T};
    std::vector<NDArray*> outputs = {dLdLohaB1};
    mmulOp.execute(inputs, outputs);
    delete lohaA1T;
  }

  // dLdA2 = B2^T @ gradProd2
  {
    auto lohaB2T = new NDArray(lohaB2->transpose());
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {lohaB2T, gradProd2};
    std::vector<NDArray*> outputs = {dLdLohaA2};
    mmulOp.execute(inputs, outputs);
    delete lohaB2T;
  }

  // dLdB2 = gradProd2 @ A2^T
  {
    auto lohaA2T = new NDArray(lohaA2->transpose());
    sd::ops::matmul mmulOp;
    std::vector<NDArray*> inputs = {gradProd2, lohaA2T};
    std::vector<NDArray*> outputs = {dLdLohaB2};
    mmulOp.execute(inputs, outputs);
    delete lohaA2T;
  }

  delete prod1;
  delete prod2;
  delete lohaDelta;
  delete dLdOutTimesInput;
  delete gradProd1;
  delete gradProd2;

  return Status::OK;
}

DECLARE_SHAPE_FN(loha_matmul_bp) {
  return SHAPELIST(
      CONSTANT(inputShape->at(0)),
      CONSTANT(inputShape->at(1)),
      CONSTANT(inputShape->at(2)),
      CONSTANT(inputShape->at(3)),
      CONSTANT(inputShape->at(4)),
      CONSTANT(inputShape->at(5))
  );
}

DECLARE_TYPES(loha_matmul_bp) {
  getOpDescriptor()->setAllowedInputTypes(ANY)->setAllowedOutputTypes({ALL_FLOATS});
}

}  // namespace ops
}  // namespace sd

#endif

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/LohaMatMul.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.transforms.custom;

import lombok.NoArgsConstructor;
import lombok.NonNull;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.common.base.Preconditions;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * LoHa (Low-Rank Hadamard Product) fused matrix multiplication operation.
 * <p>
 * Computes: output = input @ weight^T + scaling * input @ ((B1 @ A1) ⊙ (B2 @ A2))^T
 * <p>
 * where ⊙ is the Hadamard (element-wise) product. LoHa uses two low-rank decompositions
 * whose Hadamard product can represent higher effective ranks. With dimension d,
 * the effective rank can be up to d².
 * <p>
 * Input shapes:
 * <ul>
 *   <li>input: [batch, in_features]</li>
 *   <li>weight: [out_features, in_features] (frozen base weight)</li>
 *   <li>lohaA1: [dim, in_features] (trainable)</li>
 *   <li>lohaB1: [out_features, dim] (trainable)</li>
 *   <li>lohaA2: [dim, in_features] (trainable)</li>
 *   <li>lohaB2: [out_features, dim] (trainable)</li>
 * </ul>
 * <p>
 * Output shape: [batch, out_features]
 *
 * @author Adam Gibson
 * @see LoraMatMul
 */
@NoArgsConstructor
public class LohaMatMul extends DynamicCustomOp {

    private double scaling = 1.0;
    private boolean transposeWeight = true;
    private double dropout = 0.0;

    public LohaMatMul(@NonNull SameDiff sameDiff, @NonNull SDVariable input,
                      @NonNull SDVariable weight, @NonNull SDVariable lohaA1,
                      @NonNull SDVariable lohaB1, @NonNull SDVariable lohaA2,
                      @NonNull SDVariable lohaB2, double scaling) {
        super(null, sameDiff, new SDVariable[]{input, weight, lohaA1, lohaB1, lohaA2, lohaB2}, false);
        this.scaling = scaling;
        addTArgument(scaling, dropout);
        addBArgument(transposeWeight);
    }

    public LohaMatMul(@NonNull SameDiff sameDiff, @NonNull SDVariable input,
                      @NonNull SDVariable weight, @NonNull SDVariable lohaA1,
                      @NonNull SDVariable lohaB1, @NonNull SDVariable lohaA2,
                      @NonNull SDVariable lohaB2, double scaling, double dropout) {
        super(null, sameDiff, new SDVariable[]{input, weight, lohaA1, lohaB1, lohaA2, lohaB2}, false);
        this.scaling = scaling;
        this.dropout = dropout;
        addTArgument(scaling, dropout);
        addBArgument(transposeWeight);
    }

    public LohaMatMul(@NonNull SameDiff sameDiff, @NonNull SDVariable input,
                      @NonNull SDVariable weight, @NonNull SDVariable lohaA1,
                      @NonNull SDVariable lohaB1, @NonNull SDVariable lohaA2,
                      @NonNull SDVariable lohaB2, double scaling, boolean transposeWeight) {
        super(null, sameDiff, new SDVariable[]{input, weight, lohaA1, lohaB1, lohaA2, lohaB2}, false);
        this.scaling = scaling;
        this.transposeWeight = transposeWeight;
        addTArgument(scaling, dropout);
        addBArgument(transposeWeight);
    }

    public LohaMatMul(@NonNull INDArray input, @NonNull INDArray weight,
                      @NonNull INDArray lohaA1, @NonNull INDArray lohaB1,
                      @NonNull INDArray lohaA2, @NonNull INDArray lohaB2,
                      double scaling, boolean transposeWeight) {
        super(null, new INDArray[]{input, weight, lohaA1, lohaB1, lohaA2, lohaB2}, null);
        this.scaling = scaling;
        this.transposeWeight = transposeWeight;
        addTArgument(scaling, dropout);
        addBArgument(transposeWeight);
    }

    public LohaMatMul(@NonNull INDArray input, @NonNull INDArray weight,
                      @NonNull INDArray lohaA1, @NonNull INDArray lohaB1,
                      @NonNull INDArray lohaA2, @NonNull INDArray lohaB2,
                      double scaling, INDArray output) {
        super(null, new INDArray[]{input, weight, lohaA1, lohaB1, lohaA2, lohaB2},
              output == null ? null : new INDArray[]{output});
        this.scaling = scaling;
        addTArgument(scaling, dropout);
        addBArgument(transposeWeight);
    }

    @Override
    public String opName() {
        return "loha_matmul";
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> gradients) {
        SDVariable grad = gradients.get(0); // [batch, out_features]

        SDVariable input = arg(0);   // [batch, in_features]
        SDVariable weight = arg(1);  // [out_features, in_features]
        SDVariable lohaA1 = arg(2);  // [dim, in_features]
        SDVariable lohaB1 = arg(3);  // [out_features, dim]
        SDVariable lohaA2 = arg(4);  // [dim, in_features]
        SDVariable lohaB2 = arg(5);  // [out_features, dim]

        // Compute intermediate products for gradients
        // prod1 = B1 @ A1: [out_features, in_features]
        // prod2 = B2 @ A2: [out_features, in_features]
        SDVariable prod1 = sameDiff.mmul(lohaB1, lohaA1);
        SDVariable prod2 = sameDiff.mmul(lohaB2, lohaA2);

        // lohaDelta = prod1 * prod2 (Hadamard product)
        SDVariable lohaDelta = prod1.mul(prod2);

        // Gradient w.r.t. input:
        // grad_input = grad @ weight + scaling * grad @ lohaDelta
        SDVariable gradInputBase = sameDiff.mmul(grad, weight);
        SDVariable gradInputLoha = sameDiff.mmul(grad, lohaDelta).mul(scaling);
        SDVariable gradInput = gradInputBase.add(gradInputLoha);

        // Gradient w.r.t. weight (frozen)
        SDVariable gradWeight = sameDiff.zerosLike(weight);

        // For LoHa, the delta is (B1 @ A1) ⊙ (B2 @ A2)
        // Gradient flows through both decompositions

        // Gradient w.r.t. A1, B1 (through prod1, with prod2 as coefficient)
        // d/dA1: scaling * (grad @ (delta * prod2))^T @ ... but delta uses prod1
        // Actually: d(output)/d(prod1) = scaling * prod2 (element-wise)
        // Then d(prod1)/dA1 and d(prod1)/dB1

        // Simplified gradient computation using chain rule:
        // gradProd1 = scaling * grad^T @ input @ A1^T ⊙ prod2 for B1
        // This is an approximation - full gradient is more complex

        SDVariable inputTimesGrad = sameDiff.mmul(sameDiff.transpose(grad), input); // [out, in]
        SDVariable gradProd1Contribution = inputTimesGrad.mul(prod2).mul(scaling);
        SDVariable gradProd2Contribution = inputTimesGrad.mul(prod1).mul(scaling);

        // Gradient w.r.t. A1: B1^T @ gradProd1
        SDVariable gradA1 = sameDiff.mmul(sameDiff.transpose(lohaB1), gradProd1Contribution);

        // Gradient w.r.t. B1: gradProd1 @ A1^T
        SDVariable gradB1 = sameDiff.mmul(gradProd1Contribution, sameDiff.transpose(lohaA1));

        // Gradient w.r.t. A2: B2^T @ gradProd2
        SDVariable gradA2 = sameDiff.mmul(sameDiff.transpose(lohaB2), gradProd2Contribution);

        // Gradient w.r.t. B2: gradProd2 @ A2^T
        SDVariable gradB2 = sameDiff.mmul(gradProd2Contribution, sameDiff.transpose(lohaA2));

        return Arrays.asList(gradInput, gradWeight, gradA1, gradB1, gradA2, gradB2);
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes) {
        Preconditions.checkState(inputDataTypes != null && inputDataTypes.size() == 6,
            "Expected 6 input data types, got %s", inputDataTypes);
        DataType outType = inputDataTypes.get(0);
        for (int i = 1; i < inputDataTypes.size(); i++) {
            if (inputDataTypes.get(i) == DataType.DOUBLE) {
                outType = DataType.DOUBLE;
                break;
            } else if (inputDataTypes.get(i) == DataType.FLOAT && outType != DataType.DOUBLE) {
                outType = DataType.FLOAT;
            }
        }
        return Collections.singletonList(outType);
    }

    @Override
    public int getNumOutputs() {
        return 1;
    }

    @Override
    public String tensorflowName() {
        return "LohaMatMul";
    }

    @Override
    public String onnxName() {
        return "LohaMatMul";
    }
}
